popup-app/index.js

```diff
@@ -19,11 +23,45 @@ const idps = [
   }
 ]
 
-ReactDOM.render(
-  window.location.hash ? (
-    <IdpCallback afterLoggedIn={() => setTimeout(window.close, 750)} />
-  ) : (
-    <IdpSelect idps={idps} />
-  ),
-  document.getElementById('app-container')
-)
+requestAppOrigin().then(appOrigin => {
+  ReactDOM.render(
+    window.location.hash ? (
+      <IdpCallback
+        appOrigin={appOrigin}
+        afterLoggedIn={() => setTimeout(window.close, 750)}
+      />
+    ) : (
+      <IdpSelect
+        idps={idps}
+        appOrigin={appOrigin}
+        appName={process.env.TRUSTED_APP_NAME}
+      />
+    ),
+    document.getElementById('app-container')
+  )
+})
+
+async function requestAppOrigin() {
+  let appOrigin = await getStoredAppOrigin()
+  if (appOrigin) {
+    return appOrigin
+  }
+  const request = client(window.opener, '*')
+  appOrigin = await request({
+    method: 'getAppOrigin',
+    args: []
+  })
+  storeAppOrigin(appOrigin)
+  return appOrigin
```

Comment:
The `appOrigin` var is not necessary, given that `storeAppOrigin` already returns it.